packages/block-library/src/social-link/index.php

```diff
@@ -278,6 +278,8 @@ function block_core_social_link_services( $service = '', $field = '' ) {
 		),
 	);
 
+	$services_data = apply_filters( 'block_editor_social_link_services', $services_data );
+
```

Comment:
I'm not in love with:

* Adding a block-specific API in our core library
* The fact that this API exposes implementation details (`$services_data`) of that block's rendering logic

I'd love to find a solution that could inform other multi-variation server-rendered blocks, or at least a solution that is more consistent with how we already extend or intercept blocks. In the absence of that, I almost prefer that third parties just use the standard `render_block_core/social-link` filter to intercept a block that matches their variation and swap out the serialised SVG.

<details><summary>Example</summary><pre><code>
add_filter(
	'render_block_core/social-link',
	function ( $block_content, $parsed_block ) {
		if ( 'my-variation' === $parsed_block['attrs']['service'] ) {
			preg_replace( ... );
		}
		return $block_content;
	},
	10,
	2
);
</code></pre></details>

I'm not sure. What do you think, @gziolo?

Reply:
Some thoughts I have on this topic.

First of all, I see that we use `block_core_social_link_services` to keep the list of all available services on the server and we keep a very similar list of services as an array of variations on the client. They are mostly the same but keep the same data in different formats. The biggest difference present is for icons that are React elements from `@wordpress/icons` on the client and regular SVGs on the server.

In https://github.com/WordPress/gutenberg/pull/29095, @gwwar added block variations for the Navigation Link block on the server. We had a few follow-up tasks that improve the integration. One missing bit is the support in `block.json` file for block variations which might not work at all here. My initial reaction was that we could explore if we could do all the filtering with PHP on the server with `register_block_type_args` (or even better we add `register_block_type_{$setting_name}`) to register new block variations. However it doesn't solve the issue with icons.

Knowing all that, it is quite interesting approach to make `block_core_social_link_services` extensible despite the fact that it introduces this level of indirection through another filter that has to be documented. It might be simpler to go this route than to teach folks how to replace some pattern in the `render_block` filter. 

I don't think I'm helping much here. I wanted also to emphasize that we need to improve support for block icons that has to be defined on the client since we started using `@wordpress/icons`. It was also raised by @gwwar recently in the context of the Navigation Link block in https://github.com/WordPress/gutenberg/issues/30739#issuecomment-817908133:

> Sidenote: To fully remove the hooks.js file from navigation-link, 🤔 thinking through icons would be another one. One straightforward approach would be passing a string matching the icon name from the package. Is there any background on why we may have avoided this?

Reply:
Thanks for weighing in.

> Knowing all that, it is quite interesting approach to make `block_core_social_link_services` extensible despite the fact that it introduces this level of indirection through another filter that has to be documented. It might be simpler to go this route than to teach folks how to replace some pattern in the `render_block` filter.

As long as we're aware of the tradeoffs, that's good enough for me. @mkaz: if I recall correctly, back when Social Icons were introduced, we went with SVGs because Internet Explorer didn't support the alternative — is that correct? If so, what was the alternative? I ask this because that could inform our mid-term solutions for letting third parties provide new icons. In other words, once we [rid ourselves of the IE burden](https://make.wordpress.org/core/2021/03/25/discussion-summary-dropping-support-for-ie11/), can we come up with better APIs?

Reply:
It's a bit far back now, but I think we actually decided to treat IE(11) in a progressive-enhancement way. The conversation, if I recall correctly, was about easily recolored SVGs, and I think we found a way to show the same SVGs to IE11, just with reduced or no colors. 

SVG is still ideal to me, as they are easy to make, they are easy to colorize, and you can even do fancy animation stuff with them.

If we were to put on our dreamer-hats and relieve ourselves of even more burdens, we should look at improving caching of the SVGs by having one or more external files that are reused across pages, rather than outputting the entire SVG inline as we do now. All the while, the SVG still has to be easily recolored.

The obvious method for sprite work is SVG use, which is essentially that, and this article explains it well: https://css-tricks.com/svg-use-external-source/

The primary downside with that is that as an external sprite, extensibility might be tricky.

An alternative is to use `mask-image`, as demonstrated in this codepen: https://codepen.io/joen/pen/poRwpPj

The benefits are external files, easy recolorability, pretty simple code, and should work well for extensibility. The downside is browser support (though we should be [reasonably good](https://caniuse.com/css-masks) if we discount IE11), and the fact that you can only have single-color icons, because the icon gets "flattened" when used as a mask shape. That last tradeoff seems fine to me for social links, even if that does post a problem for the Snapchat logo:

<img width="112" alt="Screenshot 2021-04-22 at 19 07 05" src="https://user-images.githubusercontent.com/1204802/115755726-f2955d00-a39d-11eb-8e21-800ccb53e9fa.png">

Reply:
My previous comment was in response to Marcus; GitHub hadn't yet refreshed, so I just saw Joen's comment now.

Anyway, this is pretty cool:

> An alternative is to use mask-image, as demonstrated in this codepen: https://codepen.io/joen/pen/poRwpPj